File: Algorithm/Other/leetcode/59. Spiral Matrix II/GPT/Spiral-Matrix.py
# * Python 3.11.4
# * `class Solution` 形式（LeetCode準拠）
# * Pylance の型チェック対応
# * 関数単位で処理を分離
# * 処理時間・メモリ消費を最適化（O(n²)、in-placeな追加のみ）

# ```python
from typing import List

class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        """
        螺旋状に 1 から n^2 までを埋め込んだ n x n の行列を生成する
        
        Args:
            n (int): 行列のサイズ (1 <= n <= 20)
        
        Returns:
            List[List[int]]: 螺旋状に数字を埋め込んだ n x n の2次元リスト
        """
        # 結果用の n x n 行列を 0 で初期化
        matrix: List[List[int]] = [[0] * n for _ in range(n)]
        
        # 境界の初期化
        left: int = 0
        right: int = n - 1
        top: int = 0
        bottom: int = n - 1
        
        # 埋め込む数字の開始値
        num: int = 1
        target: int = n * n
        
        # 螺旋状に埋めるループ
        while num <= target:
            # 上端の行を左→右に埋める
            for col in range(left, right + 1):
                matrix[top][col] = num
                num += 1
            top += 1
            
            # 右端の列を上→下に埋める
            for row in range(top, bottom + 1):
                matrix[row][right] = num
                num += 1
            right -= 1
            
            # 下端の行を右→左に埋める
            if top <= bottom:
                for col in range(right, left - 1, -1):
                    matrix[bottom][col] = num
                    num += 1
                bottom -= 1
            
            # 左端の列を下→上に埋める
            if left <= right:
                for row in range(bottom, top - 1, -1):
                    matrix[row][left] = num
                    num += 1
                left += 1
        
        return matrix
# ```

# ---

# ### ✅ 設計上のポイント

# * **計算量**

#   * 各要素を一度だけ埋めるため **O(n²)**
# * **メモリ消費量**

#   * 結果の行列 `O(n²)` のみ。追加の配列は使わない。
# * **Pylance 型チェック**

#   * `n: int` と `List[List[int]]` を明示。

# ---

# 必要なら、次に **各処理の図解付き解析** を作って、
# `n=3` の例で「左から右 → 上から下 → 右から左 → 下から上」の順番を図解して解説できます。

File: Algorithm/Other/leetcode/54. Spiral Matrix/Claude/Spiral-Matrix.js
// この問題は行列を螺旋状に読み取る問題ですね。効率的な解法を作成します。この解法の特徴：

// **時間計算量:** O(m × n) - 各要素を一度だけ訪問
// **空間計算量:** O(1) - 結果配列以外の追加スペースは境界変数のみ

// **アルゴリズムの説明:**
// 1. 4つの境界（top, bottom, left, right）を設定
// 2. 螺旋状に以下の順序で要素を読み取り：
//    - 上の行：左から右
//    - 右の列：上から下
//    - 下の行：右から左（行が残っている場合のみ）
//    - 左の列：下から上（列が残っている場合のみ）
// 3. 各方向の処理後に対応する境界を更新
// 4. すべての要素を処理するまで繰り返し

// **最適化のポイント:**
// - 不要な条件チェックを最小化
// - 境界条件で無駄な処理を回避
// - 単一のwhile ループで効率的に処理

// この実装はLeetCodeでの時間・メモリ効率を考慮した最適解です。

/**
 * 行列を螺旋状の順序で読み取り、要素を配列として返す
 * @param {number[][]} matrix - m x n の整数行列
 * @return {number[]} - 螺旋状に読み取った要素の配列
 */
var spiralOrder = function (matrix) {
    // 空の行列チェック
    if (!matrix || matrix.length === 0 || matrix[0].length === 0) {
        return [];
    }

    const m = matrix.length;
    const n = matrix[0].length;
    const result = [];

    // 境界を定義
    let top = 0;
    let bottom = m - 1;
    let left = 0;
    let right = n - 1;

    while (top <= bottom && left <= right) {
        // 上の行を左から右へ
        for (let j = left; j <= right; j++) {
            result.push(matrix[top][j]);
        }
        top++;

        // 右の列を上から下へ
        for (let i = top; i <= bottom; i++) {
            result.push(matrix[i][right]);
        }
        right--;

        // 下の行を右から左へ（残りの行がある場合）
        if (top <= bottom) {
            for (let j = right; j >= left; j--) {
                result.push(matrix[bottom][j]);
            }
            bottom--;
        }

        // 左の列を下から上へ（残りの列がある場合）
        if (left <= right) {
            for (let i = bottom; i >= top; i--) {
                result.push(matrix[i][left]);
            }
            left++;
        }
    }

    return result;
};
